packages/main/src/plugin/kubernetes/contexts-manager-experimental.ts

```diff
@@ -61,29 +68,77 @@ export class ContextsManagerExperimental {
   }
 
   private async onAdd(event: DispatcherEvent): Promise<void> {
-    const previous = this.#healthCheckers.get(event.contextName);
-    previous?.dispose();
-    const newHC = new ContextHealthChecker(event.config);
-    newHC.onStateChange(this.onStateChange.bind(this));
-    this.#healthCheckers.set(event.contextName, newHC);
-    await newHC.start({ timeout: HEALTH_CHECK_TIMEOUT_MS });
+    // register and start health checker
+    const previousHealthChecker = this.#healthCheckers.get(event.contextName);
+    previousHealthChecker?.dispose();
+    const newHealthChecker = new ContextHealthChecker(event.config);
+    this.#healthCheckers.set(event.contextName, newHealthChecker);
+    newHealthChecker.onStateChange(this.onStateChange.bind(this));
+
+    newHealthChecker.onReachable(async (state: ContextHealthState) => {
+      // register and start permissions checker
+      const previousPermissionsChecker = this.#permissionsCheckers.get(state.contextName);
+      previousPermissionsChecker?.dispose();
+
+      const namespace = state.kubeConfig.getNamespace();
+      const newPermissionChecker = new ContextPermissionsChecker(state.kubeConfig, {
+        attrs: {
+          namespace,
+          group: '*',
+          resource: '*',
+          verb: 'watch',
+        },
+        resources: ['pods', 'deployments'],
+        onDenyRequests: [
+          {
+            attrs: {
+              namespace,
+              verb: 'watch',
+              resource: 'pods',
+            },
+            resources: ['pods'],
+          },
+          {
+            attrs: {
+              namespace,
+              verb: 'watch',
+              group: 'apps',
+              resource: 'deployments',
+            },
+            resources: ['deployments'],
+          },
+        ],
+      });
+      this.#permissionsCheckers.set(state.contextName, newPermissionChecker);
+      newPermissionChecker.onPermissionResult(this.onPermissionResult.bind(this));
+      await newPermissionChecker.start();
+    });
+
+    await newHealthChecker.start({ timeout: HEALTH_CHECK_TIMEOUT_MS });
   }
 
   private async onUpdate(event: DispatcherEvent): Promise<void> {
-    // we don't try to update the health checker, we recreate it
+    // we don't try to update the checkers, we recreate them
     return this.onAdd(event);
   }
 
   private onDelete(state: DispatcherEvent): void {
-    const previous = this.#healthCheckers.get(state.contextName);
-    previous?.dispose();
+    const previousHealthChecker = this.#healthCheckers.get(state.contextName);
+    previousHealthChecker?.dispose();
     this.#healthCheckers.delete(state.contextName);
+    const previousPermissionsChecker = this.#permissionsCheckers.get(state.contextName);
+    previousPermissionsChecker?.dispose();
+    this.#permissionsCheckers.delete(state.contextName);
```

Comment:
why using the wording previous, you are deleting stuff, they are not yet deleted you are deleting them, why previous?

Reply:
good point, I'll remove the prefix